src/proposal.mligo

```diff
@@ -136,29 +136,25 @@ let add_proposal (propose_params, period, store : propose_params * period * stor
 
 let submit_vote (proposal, vote_param, author, period, store : proposal * vote_param * address * period * storage): storage =
   let proposal_key = vote_param.proposal_key in
-
-  let proposal =
-    if vote_param.vote_type
-      then { proposal with upvotes = proposal.upvotes + vote_param.vote_amount }
-      else { proposal with downvotes = proposal.downvotes + vote_param.vote_amount }
-    in
-  let voter =
-    { voter_address = author
-    ; vote_amount = vote_param.vote_amount
-    ; vote_type = vote_param.vote_type
-    } in
   let proposal =
-    { proposal with
-      voters = voter :: proposal.voters
-    } in
+        let map_key = (author, vote_param.vote_type) in
+        let vote_item = match Map.find_opt map_key proposal.voters with
+          | Some vote_item -> { vote_item with vote_amount = vote_item.vote_amount + vote_param.vote_amount }
+          | None -> { vote_amount = vote_param.vote_amount } in
+        let voters = Map.add map_key vote_item proposal.voters
+        in match vote_param.vote_type with
+          | true -> { proposal with upvotes = vote_item.vote_amount; voters = voters  }
+          | false -> { proposal with downvotes = vote_item.vote_amount; voters = voters  } in
```

Comment:
A couple of things about this:
1. setting both `voters` and `upvotes`/`downvotes` at the same time is actually more expensive (the code to set `voters` is repeated twice.
2. having this all indented under the `let proposal =` is less readable IMO
3. possibly `vote_item` would be easier to use if it wasn't a record (I don't see the advantage of it being one here, but I'll let you decide).
4. the most important part is that `vote_amount` needs to be added to `upvotes`/`downvotes` not replace it!
    It looks like we don't have a test for this, please add one

This is my suggestion here:
```ocaml
  let voter_map_key = (author, vote_param.vote_type) in
  let vote_item = match Map.find_opt voter_map_key proposal.voters with
    | Some vote_item -> { vote_item with vote_amount = vote_item.vote_amount + vote_param.vote_amount }
    | None -> { vote_amount = vote_param.vote_amount } in
  let voters = Map.add voter_map_key vote_item proposal.voters in
  let proposal = { proposal with voters = voters } in
  let proposal =
    if vote_param.vote_type
    then { proposal with upvotes = proposal.upvotes + vote_param.vote_amount }
    else { proposal with downvotes = proposal.downvotes + vote_param.vote_amount }
  in
```

Reply:
Sorry about that mixup. 

I have made all the suggestions. It is not exactly as you suggested, because I have ditched `vote_item` and just use `nat`, as it does not made much sense to use the record (It might have added a bit more safety to the `do_unfreeze` call a little down, but it is only one place, so it is not probably a big deal).

Reply:
Oh, have updated it btw.